src/main/java/co/cask/gcp/speech/SpeechToTextTransform.java

```diff
@@ -0,0 +1,332 @@
+      if (fieldType != Schema.Type.BYTES) {
+        throw new IllegalArgumentException(String.format("Field '%s' must be of type 'bytes' but is of type " +
+                                                           "'%s'.", field.getName(), fieldType));
+      }
+    }
+  }
+
+  @Override
+  public void initialize(TransformContext context) throws Exception {
+    super.initialize(context);
+    schema = config.getSchema();
+    setRecognitionConfig();
+    speech = SpeechClient.create(getSettings());
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    ByteString audioBytes = ByteString.copyFrom((byte[]) input.get(config.audioField));
+    RecognitionAudio audio = RecognitionAudio.newBuilder()
+      .setContent(audioBytes)
+      .build();
+
+    RecognizeResponse response = speech.recognize(recognitionConfig, audio);
+    List<SpeechRecognitionResult> results = response.getResultsList();
+    Schema outputRecordSchema = getSchema(schema, config.getTranscriptionPartsField(),
+                                          config.getTranscriptionTextField());
+
+    StructuredRecord.Builder outputBuilder = StructuredRecord.builder(outputRecordSchema);
+
+    List<StructuredRecord> transcriptParts = new ArrayList<>();
+    StringBuilder completeTranscript = new StringBuilder();
+    for (SpeechRecognitionResult result : results) {
+      if (config.getTranscriptionPartsField() != null && !config.getTranscriptionPartsField().isEmpty()) {
+        addTranscriptWithConfidence(outputRecordSchema, transcriptParts, result.getAlternativesList());
+      }
+      if (config.getTranscriptionTextField() != null && !config.getTranscriptionTextField().isEmpty()) {
+        completeTranscript.append(result.getAlternativesList().get(0).getTranscript());
+      }
+    }
+    if (config.getTranscriptionPartsField() != null && !config.getTranscriptionPartsField().isEmpty()) {
+      outputBuilder.set(config.getTranscriptionPartsField(), transcriptParts);
+    }
+    if (config.getTranscriptionTextField() != null && !config.getTranscriptionTextField().isEmpty()) {
+      outputBuilder.set(config.getTranscriptionTextField(), completeTranscript.toString());
+    }
+
+    // copy other schema field values
+    List<Schema.Field> fields = input.getSchema().getFields();
+    if (fields != null) {
+      for (Schema.Field field : fields) {
+        String name = field.getName();
+        if (outputRecordSchema.getField(name) != null) {
+          outputBuilder.set(name, input.get(name));
+        }
+      }
+    }
+    emitter.emit(outputBuilder.build());
+  }
+
+  @Override
+  public void destroy() {
+    super.destroy();
+    try {
+      speech.close();
+    } catch (Exception e) {
+      // no-op
+    }
+  }
+
+  private Schema getSchema(Schema outSchema, @Nullable String transcriptPartsField,
```

Comment:
why does this generate a Schema from an output schema? Or is this really the input schema?

Reply:
Looking at the logic, this is a really weird method. It seems like it takes as input some potential output schema but also modifies it if it is not correct.

Change it to take an input schema. The input schema should not already have a parts field or a transcription field. Those should simply be added.